### ISSUE TITLE
Merge/close GitHub PRs

### ISSUE THREAD

**Issue by nicoddemus** [ESSS/err-stash#16]
Interesting links:

* https://github.com/PyGithub/PyGithub
* https://github.com/ESSS/err-stash/blob/master/err_stash.py

**Comment by gcardozo123** [comment]
Reopening this issue because this feature was reverted. Now I'll profile to check why searching for repos and branches is so slow on pygithub.

**Comment by gcardozo123** [comment]
Guys, we need to set a minimum acceptable time for Github support on `err-stash`. I was running some tests today and building merge plans for Github branches (only) took about 40s. I think the main reasons this takes so long are: 
1) the amount of repositories ESSS has;
2) the amount of branches those repos have;
3) the amount of fields that PyGithub objects have: when we request a given repo or branch, PyGithub builds a high level representation of it.

I need to take a better look at this later, but one idea that comes to my mind is that I could contribute to PyGithub to add a way to retrieve just repositories and branches names instead of their whole high level objects.

 **tl; dr;**
How long do you guys consider acceptable to build merge plans for Github?

**Comment by nicoddemus** [comment]
>How long do you guys consider acceptable to build merge plans for Github?

About under a minute?

**Comment by gcardozo123** [comment]
As I suspected, fetching branches is what makes it slow, `alfasim-sdk` for example has more than 200 branches.
```
Total time to build repos cache:  1.5430405139923096
------> time to fetch fb-test-pull-request-branch on repo mu-repo:  0.3861992359161377
------> time to fetch fb-test-pull-request-branch on repo jenkins-job-builder:  0.3909144401550293
------> time to fetch fb-test-pull-request-branch on repo py4eng-course:  0.42553114891052246
------> time to fetch fb-test-pull-request-branch on repo conda:  0.607752799987793
------> time to fetch fb-test-pull-request-branch on repo cit:  0.3811764717102051
------> time to fetch fb-test-pull-request-branch on repo cquadpack:  0.4388711452484131
------> time to fetch fb-test-pull-request-branch on repo jenkins-webapi:  0.3691904544830322
------> time to fetch fb-test-pull-request-branch on repo puppet-yum:  0.4394698143005371
------> time to fetch fb-test-pull-request-branch on repo puppet-archive:  0.4203019142150879
------> time to fetch fb-test-pull-request-branch on repo puppetlabs-lvm:  0.4564173221588135
------> time to fetch fb-test-pull-request-branch on repo err:  0.45345020294189453
------> time to fetch fb-test-pull-request-branch on repo err-jira:  0.38529300689697266
------> time to fetch fb-test-pull-request-branch on repo puppet-python:  0.5177617073059082
------> time to fetch fb-test-pull-request-branch on repo xml_factory:  0.43181777000427246
------> time to fetch fb-test-pull-request-branch on repo datatree:  0.3676164150238037
------> time to fetch fb-test-pull-request-branch on repo conda-build:  0.40647315979003906
------> time to fetch fb-test-pull-request-branch on repo conda-env:  0.35542845726013184
------> time to fetch fb-test-pull-request-branch on repo puppet-network:  0.4285922050476074
------> time to fetch fb-test-pull-request-branch on repo pyboost_ipc:  0.4191715717315674
------> time to fetch fb-test-pull-request-branch on repo jobs_done10:  0.4135746955871582
------> time to fetch fb-test-pull-request-branch on repo choco-specs:  0.3632674217224121
------> time to fetch fb-test-pull-request-branch on repo numpy:  0.38431239128112793
------> time to fetch fb-test-pull-request-branch on repo scipy:  0.3845961093902588
------> time to fetch fb-test-pull-request-branch on repo PyTables:  0.42626166343688965
------> time to fetch fb-test-pull-request-branch on repo faulthandler:  0.38085365295410156
------> time to fetch fb-test-pull-request-branch on repo esss_fix_format:  0.3705716133117676
------> time to fetch fb-test-pull-request-branch on repo isort:  0.4017481803894043
------> time to fetch fb-test-pull-request-branch on repo qt_style_sheet_inspector:  0.37871289253234863
------> time to fetch fb-test-pull-request-branch on repo conda-execute:  0.35935020446777344
------> time to fetch fb-test-pull-request-branch on repo xorg-recipes:  0.377077579498291
------> time to fetch fb-test-pull-request-branch on repo conda-devenv:  0.39213037490844727
------> time to fetch fb-test-pull-request-branch on repo steve:  0.5274782180786133
------> time to fetch fb-test-pull-request-branch on repo qmxgraph:  3.3537981510162354
------> time to fetch fb-test-pull-request-branch on repo clcache-builder:  0.38832664489746094
------> time to fetch fb-test-pull-request-branch on repo onpremise:  0.3920602798461914
------> time to fetch fb-test-pull-request-branch on repo sentry-zendesk:  0.46045947074890137
------> time to fetch fb-test-pull-request-branch on repo sentry:  2.032351493835449
------> time to fetch fb-test-pull-request-branch on repo notebooks:  0.4019896984100342
------> time to fetch fb-test-pull-request-branch on repo kraken-macros:  0.4941368103027344
------> time to fetch fb-test-pull-request-branch on repo conda-devenv-feedstock:  0.40840601921081543
------> time to fetch fb-test-pull-request-branch on repo jmessenger:  0.3907444477081299
------> time to fetch fb-test-pull-request-branch on repo skype_bot:  0.4058496952056885
------> time to fetch fb-test-pull-request-branch on repo err-jenkins:  0.38480401039123535
------> time to fetch fb-test-pull-request-branch on repo deps:  0.4130594730377197
------> time to fetch fb-test-pull-request-branch on repo conda-talk:  0.40208888053894043
------> time to fetch fb-test-pull-request-branch on repo err-stash:  0.37758541107177734
------> time to fetch fb-test-pull-request-branch on repo flask-restalchemy:  1.7071130275726318
------> time to fetch fb-test-pull-request-branch on repo ng-xform:  0.41872429847717285
------> time to fetch fb-test-pull-request-branch on repo pytest-replay:  0.48296594619750977
------> time to fetch fb-test-pull-request-branch on repo module-renamer:  0.41327834129333496
------> time to fetch fb-test-pull-request-branch on repo pytest-flask-sqlalchemy:  0.4459524154663086
------> time to fetch fb-test-pull-request-branch on repo cookiecutter-esss-pypackage:  0.4130096435546875
------> time to fetch fb-test-pull-request-branch on repo barril:  0.38924169540405273
------> time to fetch fb-test-pull-request-branch on repo lid_driven_cavity_problem:  0.3765981197357178
------> time to fetch fb-test-pull-request-branch on repo glinfo:  0.3987739086151123
------> time to fetch fb-test-pull-request-branch on repo PyDev.Formatter:  0.39899277687072754
------> time to fetch fb-test-pull-request-branch on repo hookman:  1.721578598022461
------> time to fetch fb-test-pull-request-branch on repo pytest-regressions:  0.4738774299621582
------> time to fetch fb-test-pull-request-branch on repo staged-recipes:  0.43962764739990234
------> time to fetch fb-test-pull-request-branch on repo alfasim-sdk:  2.159053087234497
------> time to fetch fb-test-pull-request-branch on repo bung:  0.4014716148376465
------> time to fetch fb-test-pull-request-branch on repo awesome-dev-articles:  0.4157445430755615
------> time to fetch fb-test-pull-request-branch on repo reaktoro:  0.862553596496582
------> time to fetch fb-test-pull-request-branch on repo primeng:  0.3977499008178711
------> time to fetch fb-test-pull-request-branch on repo serialchemy:  1.1094856262207031
------> time to fetch fb-test-pull-request-branch on repo oop-ext:  0.44851160049438477
------> time to fetch fb-test-pull-request-branch on repo jira2latex:  0.23646163940429688
----------------------------->>> TOTAL FETCH BRANCHES TIME:  37.66378593444824
----------------------------->>> TOTAL FETCH PRS TIME:  0.26371264457702637
```

**Comment by nicoddemus** [comment]
Hmm OK, I also did look at the GitHub API docs and I don't see a way to "find" branches, only to list them all or access them directly by name.

How about we approach this differently: we first find the unambiguous branch name on our internal Stash server. Once we find it (we would abort anyway if the user-entered query is ambiguous), then we can access the branch on GitHub repos directly by name. We will be using our bot mainly when mixing GitHub and Stash branches anyway, and this avoids having to list all branches.

**Comment by gcardozo123** [comment]
Sounds good to me, then it would be slow (I expect <1min) only when working exclusively on Github branches.

**Comment by nicoddemus** [comment]
Agreed, but even then we don't usually use Bender for those merges, so we should be fine.

**Comment by gcardozo123** [comment]
The PR is not ready yet, but once it is, is there a way for us to test my `err-stash` branch on ESSS before merging it into `master`? I'm concerned because the last time I ended up having to revert my merge.

**Comment by nicoddemus** [comment]
There are instructions on how to do that locally on the readme IIRC

**Comment by gcardozo123** [comment]
I tried your suggestion of passing to `GithubAPI` the name of the branch that the `StashApi` previously found, but the performance was about the same because we still need to check all repositories for the given branch name. I even tried returning an empty list in this [line](https://github.com/ESSS/err-stash/blob/220cf80a1c7e01b94630d474052d976031574a2b/err_stash.py#L93) but there was no considerable gains. :(

**Comment by nicoddemus** [comment]
Hmmm something is not right:

```python
import time

from github import Github, GithubException

token = 'XXXX'
github = Github(login_or_token=token)
start = time.perf_counter()
for repo in github.get_organization('ESSS').get_repos():
    try:
        print(repo.get_branch('foo-bar'))
    except GithubException:
        print(f'NOT FOUND for {repo.name}')

print(f'total time: {time.perf_counter() - start}')

```

```
$ X:\python-gh-test\.env37\Scripts\python.exe X:/python-gh-test/test-branch-access.py
NOT FOUND for mu-repo
NOT FOUND for jenkins-job-builder
NOT FOUND for py4eng-course
NOT FOUND for conda
NOT FOUND for cit
NOT FOUND for cquadpack
NOT FOUND for jenkins-webapi
NOT FOUND for puppet-yum
NOT FOUND for puppet-archive
NOT FOUND for puppetlabs-lvm
NOT FOUND for err
NOT FOUND for err-jira
NOT FOUND for puppet-python
NOT FOUND for xml_factory
NOT FOUND for datatree
NOT FOUND for conda-build
NOT FOUND for conda-env
NOT FOUND for puppet-network
NOT FOUND for pyboost_ipc
NOT FOUND for jobs_done10
NOT FOUND for choco-specs
NOT FOUND for numpy
NOT FOUND for scipy
NOT FOUND for PyTables
NOT FOUND for faulthandler
NOT FOUND for esss_fix_format
NOT FOUND for isort
NOT FOUND for qt_style_sheet_inspector
NOT FOUND for conda-execute
NOT FOUND for xorg-recipes
NOT FOUND for conda-devenv
NOT FOUND for steve
NOT FOUND for qmxgraph
NOT FOUND for clcache-builder
NOT FOUND for onpremise
NOT FOUND for sentry-zendesk
NOT FOUND for sentry
NOT FOUND for notebooks
NOT FOUND for kraken-macros
NOT FOUND for conda-devenv-feedstock
NOT FOUND for jmessenger
NOT FOUND for skype_bot
NOT FOUND for err-jenkins
NOT FOUND for deps
NOT FOUND for conda-talk
NOT FOUND for err-stash
NOT FOUND for flask-restalchemy
NOT FOUND for ng-xform
NOT FOUND for pytest-replay
NOT FOUND for module-renamer
NOT FOUND for pytest-flask-sqlalchemy
NOT FOUND for cookiecutter-esss-pypackage
NOT FOUND for barril
NOT FOUND for lid_driven_cavity_problem
NOT FOUND for glinfo
NOT FOUND for PyDev.Formatter
NOT FOUND for hookman
NOT FOUND for pytest-regressions
NOT FOUND for staged-recipes
NOT FOUND for alfasim-sdk
NOT FOUND for bung
NOT FOUND for awesome-dev-articles
NOT FOUND for reaktoro
NOT FOUND for primeng
NOT FOUND for serialchemy
NOT FOUND for oop-ext
NOT FOUND for jira2latex
total time: 17.521937175

Process finished with exit code 0
```

**Comment by gcardozo123** [comment]
Did you try that using a valid Github token? Otherwise otherwise you'll quickly reach maximum number of requests and alway enter the 
```python
except GithubException:
    print(f'NOT FOUND for {repo.name}')
```

**Comment by nicoddemus** [comment]
> Did you try that using a valid Github token? Otherwise otherwise you'll quickly reach maximum number of requests and alway enter the

Good point, but yes:

```python
import time

from github import Github, GithubException

token = 'XXXXX'
github = Github(login_or_token=token)
start = time.perf_counter()
for repo in github.get_organization('ESSS').get_repos():
    try:
        print(repo.get_branch('foo-bar'))
    except GithubException as e:
        print(f'NOT FOUND for {repo.name} {e.data["message"]}')

print(f'total time: {time.perf_counter() - start}')
```

```
...
NOT FOUND for reaktoro Branch not found
NOT FOUND for primeng Branch not found
NOT FOUND for serialchemy Branch not found
...
```

Also we should clean up the forks we have there to speed things up.

**Comment by nicoddemus** [comment]
Actually we can skip forks:

```python
import time

from github import Github, GithubException

token = 'XXX'
github = Github(login_or_token=token)
start = time.perf_counter()
for repo in github.get_organization('ESSS').get_repos():
    if repo.fork:
        print(f'SKIP {repo.name} because is a fork')
        continue
    try:
        print(repo.get_branch('foo-bar'))
    except GithubException as e:
        print(f'NOT FOUND {repo.name} {e.data["message"]}')

print(f'total time: {time.perf_counter() - start}')
```

```
SKIP mu-repo because is a fork
SKIP jenkins-job-builder because is a fork
NOT FOUND py4eng-course Branch not found
SKIP conda because is a fork
SKIP cit because is a fork
NOT FOUND cquadpack Branch not found
SKIP jenkins-webapi because is a fork
SKIP puppet-yum because is a fork
SKIP puppet-archive because is a fork
SKIP puppetlabs-lvm because is a fork
SKIP err because is a fork
SKIP err-jira because is a fork
SKIP puppet-python because is a fork
NOT FOUND xml_factory Branch not found
SKIP datatree because is a fork
SKIP conda-build because is a fork
SKIP conda-env because is a fork
SKIP puppet-network because is a fork
NOT FOUND pyboost_ipc Branch not found
NOT FOUND jobs_done10 Branch not found
NOT FOUND choco-specs Branch not found
SKIP numpy because is a fork
SKIP scipy because is a fork
SKIP PyTables because is a fork
SKIP faulthandler because is a fork
NOT FOUND esss_fix_format Branch not found
SKIP isort because is a fork
NOT FOUND qt_style_sheet_inspector Branch not found
SKIP conda-execute because is a fork
NOT FOUND xorg-recipes Branch not found
NOT FOUND conda-devenv Branch not found
NOT FOUND steve Branch not found
NOT FOUND qmxgraph Branch not found
NOT FOUND clcache-builder Branch not found
SKIP onpremise because is a fork
NOT FOUND sentry-zendesk Branch not found
SKIP sentry because is a fork
NOT FOUND notebooks Branch not found
NOT FOUND kraken-macros Branch not found
SKIP conda-devenv-feedstock because is a fork
NOT FOUND jmessenger Branch not found
SKIP skype_bot because is a fork
NOT FOUND err-jenkins Branch not found
NOT FOUND deps Branch not found
NOT FOUND conda-talk Branch not found
NOT FOUND err-stash Branch not found
NOT FOUND flask-restalchemy Branch not found
NOT FOUND ng-xform Branch not found
NOT FOUND pytest-replay Branch not found
NOT FOUND module-renamer Branch not found
NOT FOUND pytest-flask-sqlalchemy Branch not found
NOT FOUND cookiecutter-esss-pypackage Branch not found
NOT FOUND barril Branch not found
SKIP lid_driven_cavity_problem because is a fork
NOT FOUND glinfo Branch not found
SKIP PyDev.Formatter because is a fork
NOT FOUND hookman Branch not found
NOT FOUND pytest-regressions Branch not found
SKIP staged-recipes because is a fork
NOT FOUND alfasim-sdk Branch not found
NOT FOUND bung Branch not found
NOT FOUND awesome-dev-articles Branch not found
SKIP reaktoro because is a fork
SKIP primeng because is a fork
NOT FOUND serialchemy Branch not found
NOT FOUND oop-ext Branch not found
NOT FOUND jira2latex Branch not found
total time: 10.692057574
```

This reduces the total time to 10s. 👍

**Comment by gcardozo123** [comment]
I didn't thought about forks, I thought everyone created branches directly on our repos. 
![Alt text](https://66.media.tumblr.com/60aeee62dc1aee0c3c0fbad1702eb860/tumblr_inline_pfp352ORsk1r4hkfd_250.png)

Alright, I'll try that later.

**Comment by nicoddemus** [comment]
> I didn't thought about forks, I thought everyone created branches directly on our repos.

That was the common practice, but we stopped doing that some time ago. 😁

**Comment by gcardozo123** [comment]
Gosh, I'm trying this again and looks like that trying to match the branch name is actually the problem (even though I tested it before and obtained different results somehow).
I changed:
```diff
- return [branch for branch in list(repo.get_branches()) if branch_name in branch.name]
+ return []
```
and then:
```
time fetching branches from  mu-repo :  0.26490159999999996
time fetching branches from  jenkins-job-builder :  0.22801699999999947
time fetching branches from  py4eng-course :  0.2387592999999999
time fetching branches from  conda :  0.22008859999999864
time fetching branches from  cit :  0.23971810000000104
time fetching branches from  cquadpack :  0.23035920000000054
time fetching branches from  jenkins-webapi :  0.2595842000000008
time fetching branches from  puppet-yum :  0.21624470000000073
time fetching branches from  puppet-archive :  0.24038749999999887
time fetching branches from  puppetlabs-lvm :  0.22362549999999892
time fetching branches from  err :  0.2185164000000004
time fetching branches from  err-jira :  0.2295183999999999
time fetching branches from  puppet-python :  0.2553453000000001
time fetching branches from  xml_factory :  0.2392903999999998
time fetching branches from  datatree :  0.23731710000000028
time fetching branches from  conda-build :  0.24887449999999944
time fetching branches from  conda-env :  0.24712659999999964
time fetching branches from  puppet-network :  0.26065249999999907
time fetching branches from  pyboost_ipc :  0.25180639999999954
time fetching branches from  jobs_done10 :  0.23941639999999964
time fetching branches from  choco-specs :  0.23959890000000073
time fetching branches from  numpy :  0.23193799999999953
time fetching branches from  scipy :  0.23455000000000048
time fetching branches from  PyTables :  0.2842412000000003
time fetching branches from  faulthandler :  0.23075150000000022
time fetching branches from  esss_fix_format :  0.23887919999999951
time fetching branches from  isort :  0.22055950000000024
time fetching branches from  qt_style_sheet_inspector :  0.22149530000000084
time fetching branches from  conda-execute :  0.24081629999999876
time fetching branches from  xorg-recipes :  0.24100900000000003
time fetching branches from  conda-devenv :  0.27510559999999984
time fetching branches from  steve :  0.2532595000000004
time fetching branches from  qmxgraph :  0.23188679999999984
time fetching branches from  clcache-builder :  0.23574780000000217
time fetching branches from  onpremise :  0.22854620000000025
time fetching branches from  sentry-zendesk :  0.2455186999999981
time fetching branches from  sentry :  0.23754579999999947
time fetching branches from  notebooks :  0.2616662000000005
time fetching branches from  kraken-macros :  0.24049309999999835
time fetching branches from  conda-devenv-feedstock :  0.2282687999999986
time fetching branches from  jmessenger :  0.23205480000000023
time fetching branches from  skype_bot :  0.22968780000000066
time fetching branches from  err-jenkins :  0.2361027
time fetching branches from  deps :  0.2596588000000004
time fetching branches from  conda-talk :  0.2444525999999989
time fetching branches from  err-stash :  0.23705220000000082
time fetching branches from  flask-restalchemy :  0.23068800000000067
time fetching branches from  ng-xform :  0.24183780000000255
time fetching branches from  pytest-replay :  0.2519127000000019
time fetching branches from  module-renamer :  0.27414170000000126
time fetching branches from  pytest-flask-sqlalchemy :  0.24221769999999765
time fetching branches from  cookiecutter-esss-pypackage :  0.26588220000000007
time fetching branches from  barril :  0.23843829999999855
time fetching branches from  lid_driven_cavity_problem :  0.24433770000000266
time fetching branches from  glinfo :  0.3076654000000012
time fetching branches from  PyDev.Formatter :  0.22077210000000136
time fetching branches from  hookman :  0.2643620999999996
time fetching branches from  pytest-regressions :  0.259089000000003
time fetching branches from  staged-recipes :  0.2641499999999972
time fetching branches from  alfasim-sdk :  0.23609940000000051
time fetching branches from  bung :  0.250275300000002
time fetching branches from  awesome-dev-articles :  0.23094619999999821
time fetching branches from  reaktoro :  0.2502196000000012
time fetching branches from  primeng :  0.2896883000000017
time fetching branches from  serialchemy :  0.22096949999999893
time fetching branches from  oop-ext :  0.24127409999999827
time fetching branches from  jira2latex :  0.23029430000000062
Total time creating Github plans:  16.3129549
```
The downside is that in cases where the branch only exists on Github the final user needs to type the full branch name, otherwise it won't find it.

**Comment by gcardozo123** [comment]
Just found another thing that's slowing down plans creation:
When I'm iterating over repos, fetching branches I do this
```python
    for repo in github_repos:
        organization = repo.owner.name
```
in order to `err-stash` to be able to work with multiple Github organizations (instead of hard coding `organization = 'ESSS'`). Turns out that `repo.owner.name` does:

```python
@property
    def owner(self):
        """
        :type: :class:`github.NamedUser.NamedUser`
        """
        self._completeIfNotSet(self._owner)
        return self._owner.value
```
which later on does more requests:
```python
    def __complete(self):
        headers, data = self._requester.requestJsonAndCheck(
            "GET",
            self._url.value
        )
        self._storeAndUseAttributes(headers, data)
        self.__completed = True
```
Looks like this makes the creation time go from ~15s to ~30s.

So I'll see a way of associating repos to their owners without having to request that from Github (as we already have access to this info).

**Comment by nicoddemus** [comment]
Using threads:

```python
import time
from concurrent.futures import ThreadPoolExecutor, as_completed

from github import Github, GithubException

token = 'X'
branch_name = 'fb-ASIM-265-perkins-valve'
github = Github(login_or_token=token)

start = time.perf_counter()
futures = {}


with ThreadPoolExecutor(max_workers=16) as executor:
    for repo in github.get_organization('ESSS').get_repos():
        f = executor.submit(repo.get_branch, branch_name)
        futures[f] = repo.name


    for f in as_completed(futures.keys()):
        repo_name = futures[f]
        try:
            print(repo_name, f.result())
        except GithubException as e:
            print(f'NOT FOUND {repo_name} {e.data["message"]}')

print(f'total time: {time.perf_counter() - start}')
``` 

```
(.env36-pytest462) λ python gh-check.py
NOT FOUND clcache-builder Branch not found
NOT FOUND xml_factory Branch not found
NOT FOUND conda-talk Branch not found
qmxgraph Branch(name="fb-ASIM-265-perkins-valve")
NOT FOUND jobs_done10 Branch not found
NOT FOUND jenkins-job-builder Branch not found
NOT FOUND conda-env Branch not found
NOT FOUND choco-specs Branch not found
NOT FOUND pyboost_ipc Branch not found
NOT FOUND conda Branch not found
NOT FOUND mu-repo Branch not found
NOT FOUND puppet-yum Branch not found
NOT FOUND py4eng-course Branch not found
NOT FOUND xorg-recipes Branch not found
NOT FOUND conda-build Branch not found
NOT FOUND jenkins-webapi Branch not found
NOT FOUND skype_bot Branch not found
NOT FOUND err-jira Branch not found
NOT FOUND puppet-network Branch not found
NOT FOUND conda-execute Branch not found
NOT FOUND puppet-archive Branch not found
NOT FOUND jmessenger Branch not found
NOT FOUND qt_style_sheet_inspector Branch not found
NOT FOUND datatree Branch not found
NOT FOUND cookiecutter-esss-pypackage Branch not found
NOT FOUND isort Branch not found
NOT FOUND cquadpack Branch not found
NOT FOUND puppetlabs-lvm Branch not found
NOT FOUND esss_fix_format Branch not found
NOT FOUND err Branch not found
NOT FOUND faulthandler Branch not found
NOT FOUND puppet-python Branch not found
NOT FOUND sentry Branch not found
NOT FOUND pytest-replay Branch not found
NOT FOUND PyTables Branch not found
NOT FOUND sentry-zendesk Branch not found
NOT FOUND ng-xform Branch not found
NOT FOUND cit Branch not found
NOT FOUND scipy Branch not found
NOT FOUND onpremise Branch not found
NOT FOUND flask-restalchemy Branch not found
NOT FOUND steve Branch not found
NOT FOUND numpy Branch not found
NOT FOUND barril Branch not found
NOT FOUND pytest-flask-sqlalchemy Branch not found
NOT FOUND module-renamer Branch not found
NOT FOUND lid_driven_cavity_problem Branch not found
NOT FOUND glinfo Branch not found
NOT FOUND PyDev.Formatter Branch not found
NOT FOUND err-stash Branch not found
NOT FOUND conda-devenv Branch not found
NOT FOUND deps Branch not found
NOT FOUND bung Branch not found
NOT FOUND alfasim-sdk Branch not found
NOT FOUND hookman Branch not found
NOT FOUND kraken-macros Branch not found
NOT FOUND notebooks Branch not found
NOT FOUND awesome-dev-articles Branch not found
NOT FOUND staged-recipes Branch not found
NOT FOUND pytest-regressions Branch not found
NOT FOUND reaktoro Branch not found
NOT FOUND primeng Branch not found
NOT FOUND err-jenkins Branch not found
NOT FOUND serialchemy Branch not found
NOT FOUND oop-ext Branch not found
NOT FOUND jira2latex Branch not found
NOT FOUND conda-devenv-feedstock Branch not found
total time: 3.9637464590262512
```

**Comment by gcardozo123** [comment]
Niiiiice!

**Comment by gcardozo123** [comment]
Do you think we should pass a `timeout` to `as_completed`?

**Comment by nicoddemus** [comment]
I don't think it is necessary, eventually the underlying connection will timeout by itself.

**Comment by gcardozo123** [comment]
Something is preventing me from setting my user name, so I can't test errbot locally. 😞 
```
[@gcardozo ➡ @errbot] >>> !plugin config Stash { 'STASH_URL': 'https://eden.esss.com.br/stash', 'STASH_PROJECTS': ['ESSS'], 'GITHUB_ORGANIZATIONS': ['ESSS'], }
Plugin configuration done.

[@gcardozo ➡ @errbot] >>> !stash token
Your API Token is: <token_here> (user: None)

[@gcardozo ➡ @errbot] >>> !github token
Your Github Token is: <token_here> (user: None)

[@gcardozo ➡ @errbot] >>> !asuser gcardozo
You are now: @gcardozo

[@gcardozo ➡ @errbot] >>> !stash token
Your API Token is: <token_here> (user: None)

[@gcardozo ➡ @errbot] >>> !github token
Your Github Token is: <token_here> (user: None)
```

**Comment by gcardozo123** [comment]
## For documentation sake

during the implementation of this feature, we cached repositories and added threads during merge plans creation. This things brought great performance improvements as pointed out by @nicoddemus [in this comment](https://github.com/ESSS/err-stash/issues/16#issuecomment-516910476).
I also tried to add these 2 improvements: repositores cache and threads to the `StashAPI`, but the performance gain was minimum (about 3s), so we decided to not include them as they bring more complexity to the implementation. 

But here it is in case we ever decide to go back in this decision:
## Repository cache in `StashAPI`:
```python
class StashAPI:
    """
    Thin access to the stashy API.

    We have this thin layer in order to mock it during testing.
    """

    def __init__(self, url, *, username, password, projects=[]):
        self._stash = stashy.connect(url, username=username, password=password)
        self._url = url

        # projects cache
        self.projects = {
            project: self._stash.projects[project] for project in projects
        }

        # repositories cache
        self.repos = dict()
        self.paginated_repos = dict()

        for project in self.projects:
            paginated_repos = self.projects[project].repos.list()
            self.paginated_repos[project] = paginated_repos

            self.repos[project] = {
                # this line looks strange but this is the way to obtain `Repository` objects
                # from Stash api
                repo['slug']: self._stash.projects[project].repos[repo['slug']] for repo in paginated_repos
            }

    @property
    def url(self):
        return self._url

    def fetch_repos(self, project):
        return self.paginated_repos[project]

    def fetch_branches(self, project, slug, filter_text):
        return self.repos[project][slug].branches(filterText=filter_text)

    def delete_branch(self, project, slug, branch):
        return self.repos[project][slug].delete_branch(branch)

    def fetch_pull_requests(self, project, slug):
        return self.repos[project][slug].pull_requests.all()

    def fetch_pull_request(self, project, slug, pr_id):
        return self.repos[project][slug].pull_requests[pr_id]

    def fetch_repo_commits(self, project, slug, until, since):
        return self.repos[project][slug].commits(until, since)

```
## StashAPI plans with threads:
```python
def create_plans(stash_api, github_api, stash_projects, github_organizations, branch_text):
    """
    Go over all the branches in all repositories searching for branches and PRs that match the given branch text.

    :rtype: List[MergePlan]
    """
    # Plans for Stash repos:
    stash_repos = []
    for project in stash_projects:
        stash_repos += stash_api.fetch_repos(project)

    plans = []
    has_prs = False
    futures = dict()
    with ThreadPoolExecutor(max_workers=30) as executor:
        for repo in stash_repos:
            slug = repo['slug']
            project = repo['project']['key']
            f = executor.submit(stash_api.fetch_branches, project, slug, branch_text)
            futures[f] = (project, slug)

        for f in as_completed(futures.keys()):
            project, slug = futures[f]
            branches = list(f.result())
            if not branches:
                continue

            plan = MergePlan(project, slug)
            plans.append(plan)
            plan.branches = branches
            branch_ids = [x['id'] for x in plan.branches]
            prs = list(stash_api.fetch_pull_requests(project, slug))
            for pr in prs:
                if pr['fromRef']['id'] in branch_ids:
                    has_prs = True
                    plan.pull_requests.append(pr)
            if plan.pull_requests:
                plan.to_branch = plan.pull_requests[0]['toRef']['id']
```
